refactor: clean up some weather & terrain stuff

DayKev opened the pull request:
## What are the changes the user will see?
- Flower Gift will work in the primal sun weather.
- The arena flyout will no longer display "none" after weather/terrain ends.

## Why am I making these changes?
Code cleanup, etc.

## What are the changes from a developer perspective?
- `Arena#hasWeather` and `Arena#hasTerrain` now use a spread parameter.
- Added `readonly` to a bunch of stuff.
- Added `Arena#canSetTerrain`.
- Added helper function `getTerrainAnim`.
- Converted `WeatherType` from `enum` to `const` object due to some overlapping changes I had already made.
- Incidentally fixed Flower Gift so that it works in the primal sun weather too.
- Fixed the arena flyout so it no longer displays "none" after weather/terrain ends.

## Checklist
- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?

DayKev committed:
Convert `WeatherType` to `const` object

DayKev committed:
Refactor `tryOverrideWeather`, `canSetWeather` and weather helper funcs

DayKev committed:
Refactor terrain helper funcs, add `getTerrainAnim` + `canSetTerrain`

DayKev committed:
Convert `Arena#hasWeather` to use spread parameter

DayKev committed:
Convert `Arena#hasTerrain` to use spread parameter

DayKev committed:
Fix name of `EffectiveStatMultiplier`, update more types

DayKev committed:
Make Cherrim form revert trigger auto-updating with new weather types

DayKev committed:
Fix arena flyout text showing "none" when weather ends

innerthunder commented on src/constants/weather-constants.ts:
```diff
@@ -4,14 +4,23 @@ import { WeatherType } from "#enums/weather-type";
 
 /** Weather types that are associated with the sunny weather. */
-export const SUNNY_WEATHER_TYPES = Object.freeze([WeatherType.SUNNY, WeatherType.HARSH_SUN]);
+export const SUNNY_WEATHER_TYPES = Object.freeze([WeatherType.SUNNY, WeatherType.HARSH_SUN] as const);
```
I notice there's two different styles being used here:

```ts
// A
Object.freeze<WeatherType[]>([...]);
```

vs

```ts
// B
Object.freeze([...] as const);
```

A seems to be used more widely outside of this file. I'm not sure if there's a specific reason B is being used, but I think it'd be better to stay consistent (not to mention that B is interpreted as a tuple).

DayKev replied:
The second style is necessary so that it can be used for spread parameters in the form of:
```ts
function weatherStuff(...params: [WeatherType, ...WeatherType[]]) {}
weatherStuff(...SUNNY_WEATHER_TYPES)
```
[TS Playground Example](https://www.typescriptlang.org/play/?target=10&module=7#code/KYDwDg9gTgLgBAYwgOwM7wOrAIYwBbBQAqAnmMHALxwDeAUHHAHIDyTAogFxwAMANAzgB6AFQjBcEXGwBrbHABEACWxRUeOAGUArsgA2ASwDmeGAolSAAqmAUalw8hlxTMMKk5ChAI217v2OQAJgbYAHS+-thGqkHAyGHIwDBCAO4GMgZCKmp4APqouoYm8DnqWkXGpnAAtHAAQn4BwaEAvhZCgpoAqkxMAJrcAIwCjKJWNnYOBk4uMG4eXpHNwCHhy9Gx8YnJaRlZAErYM3BHJ3WNUS3YrZKdjAcAggCSTNwATKPCYnDWtrTTWaudyeHxNQKrUIRcExKBxBJJFLpTJCTTYZBBdDQAC2eQAFKkcPhCHkkBiDDADCgAJRadGYmA4uB4rC4AhQOAAYRQIUpNNqDXB11uInuWkeTAAIpoiCwDgBZbgAZi+41+kwBjmcwMWYKukPWMK2CN2yKyKgMenxhLZJLJvKpyFpFr0zNZxI53PJfKdAsuKzWIrFSheABluAAWVU-P5TLVzBagjbXaFRWHwnZI-ao5AQVJaXP5i5Cg1BrqsDDcACs0Ym-3s8Z1SZLa1TAXT20RexRADEIEY4H2B8X9YG7oIeywAOLcABstfV9cB2vmIKWLahGw7JqzKKUOAAbiQ8lBjsg4PvsEfTme-RubuPGEp2I8AGr9PJPV7cADstYs0hyIo7AgDAUDANiwB6CQF6qOUOj6FUZgAbGmozCuibrqOm5GnCnamtmIFgRBUHHngcH5IUiElHARHgZB0GwbkFTUdUI4Bm0HSCCGByaEoeQ9G8cAABwLqhDboQma56hxhppsambdlkmhgSgRh5OkGKoFoqnIAOGAzJid7YQ+opdEQBxsFOeQYK80rcAAnAItzYNpZLoAA3HQdCgJAsBwDAZAUO67KkOQVDMoF5AQAAZnAIWEGFwDUgA2jIwAkLFAVBVlCXEEFAC6Xk+eA0DwO58CCQMNkvkQz4HHkRD9AACuwmh5I8EUsN4ABWwAIDAYQxeBwAAF7AAAPHlSUpQVAB8eIpdNQVhFV-R8PFRKhStPF8QJvQFdSXm+WViAoOgWi9NVGC1fVjUtW1eT1F1vX9YNw22ONi3LeQq1Xetm22vlv27fxgkFdIbnnTAR3eTFugDY6AXAOgeJhOjYCqNg2IeHAS1bYlQUbejYQ-cAs2HbQ7QwCjMBo+ja01Y8dXsA1TWte1jxHcIQhwBNNR1IQUDQHABhxdoqAzAOkH4BAQRwI8dA06jJOMzdzN3ezj31EdQA)

innerthunder replied:
The util type I mentioned earlier could help with that.

[Example](https://www.typescriptlang.org/play/?target=10&module=7#code/KYDwDg9gTgLgBAYwgOwM7wOrAIYwBbBQAqAnmMHALxwDeAUHHAHIDyTAogFxwAMANAzgB6AFQjBcEXGwBrbHABEACWxRUeOAGUArsgA2ASwDmeGAolSAAqmAUalw8hlxTMMKk5ChAI217v2OQAJgbYAHS+-thGqkHAyGHIwDBCAO4GMgZCKmp4APqouoYm8DnqWkXGpnAAtHAAQn4BwaEAvhZCgpoAqkxMAJrcAIwCjKJWNnYOBk4uMG4eXpHNwCHhy9Gx8YnJaRlZAErYM3BHJ3WNUS3YrZKdjAcAggCSTNwATKPCYnDWtrTTWaudyeHxNQKrUIRcExKBxBJJFLpTJCTTYZBBdDQAC2eQAFKkcPhCHkkBiDDADCgAJRadGYmA4uB4rC4AhQOAAYRQIUpNNqDXB11uInuWkeTAAIpoiCwDgBZbgAZi+41+kwBjmcwMWYKukPWMK2CN2yKyKgMenxhLZJLJvKpyFpFr0zNZxI53PJfKdAsuKzWIrFSheABluAAWVU-P5TLVzBagjbXaFRWHwnZI-ao5AQVJaXP5i5Cg1BrqsDDcACs0Ym-3s8Z1SZLa1TAXT20RexRADEIEY4H2B8X9YG7oIeywAOLcABstfV9cB2vmIKWLahGw7JqzKKUOAAbiQ8lBjsg4PvsEfTme-RubuPGEp2I8AGr9PJPV7cADstYs0hyIo7AgDAUDANiwB6CQF6qOUOj6FUZgAbGmozCuibrqOm5GnCnamtmIFgRBUHHngcH5IUiElHARHgZB0GwbkFTUdUI4Bm0HSCCGByaEoeQ9G8cAABwLqhDboQma56hxhppsambdlkmhgSgRh5OkGKoFoqnIAOGAzJid7YQ+opdEQBxsFOeQYK80rcAAnAItzYNpZLoAA3HQdCgJAsBwDAZAUO67KkOQVDMoF5AQAAZnAIWEGFwDUgA2jIwAkLFAVBVlCXEEFAC6Xm+dA8BRRQjwwKGODoCwSQsDFAA8RAAHwRSlRB8HAYQ9UQKUFUV3klf57nwIJAw2S+RDPgceREP0AAK7CaHkjwRSw3gAFbAAgMBhDF4HAAAXsAjWVdVrkwHVwANY1eVJS1LV4il91BWE439F1r3kGEPF8QJvQFdSxXgKViAoOgWi9BNGBTTNc2LcteT1OtW07XtB22Cdz3fcA73Q598VEqFb1-fxgkFdIbkQzAwPeTFui7Y6AXAOgeI9WEYCqNg2IeHA501Vd9VNbjLW0jQ7QwKzMDsz1H2TY803sLN81LStjx01LbMc-LsOK-DqtI-UwNAA)

DayKev replied:
Oh, apply it to the const too... hmm.

innerthunder commented on src/data/init/init-abilities.ts:
```diff
@@ -1691,9 +1671,9 @@ export function initAbilities() {
   return (_pokemon: Pokemon) => {
-    return globalScene.arena.hasTerrain([...terrainTypes]);
+    return globalScene.arena.hasTerrain(...terrainTypes);
   };
```
Idk if you prefer it being explicit, but you should be able to omit the unused parameter entirely:

```ts
return () => globalScene.arena.hasTerrain(...terrainTypes);
```

<sub>GitHub won't let me make a proper suggestion :despair:</sub>

DayKev replied:
Normally I would prefer it to be explicit but the function return type is just one line up so it's easy to find the full type. When it's buried away from an easily-hoverable type (like within the ability/move init files) then I'd keep it explicit.

innerthunder commented on src/field/arena.ts:
```diff
@@ -185,53 +186,48 @@ export class Arena {
    */
-  public hasWeather(weather: WeatherType | WeatherType[]): boolean {
-    return coerceArray(weather).includes(this.weatherType);
+  public hasWeather(...weather: readonly [WeatherType, ...WeatherType[]]): boolean {
```
I wonder if a util type would be useful for inputs like this, e.g.

```ts
export type AtLeastOne<T> = readonly [T, ...T[]];
```

innerthunder commented on src/field/arena.ts:
```diff
@@ -357,22 +350,33 @@ export class Arena {
+    return terrain.includes(this.terrainType);
+  }
+
+  /** Sets terrain to the override specified in `overrides.ts` */
```
```suggestion
  /** Sets terrain to the override specified in {@linkcode activeOverrides | overrides.ts} */
```

innerthunder commented on src/field/arena.ts:
```diff
@@ -357,22 +350,33 @@ export class Arena {
+    if (terrain === this.terrainType) {
+      return false;
+    }
     return true;
```
```suggestion
    return terrain !== this.terrainType;
```

Also consider renaming `terrain` to `terrainType`

DayKev committed:
Apply suggestions

- Simplify `Arena#canSetTerrain`

- Update docs for `Arena#tryOverrideTerrain`

- Simplify `getTerrainCondition`

DayKev committed:
Add new utility type `AtLeastOneArray<T>`

This type requires an array to contain at least one item

DayKev committed:
Implement `AtLeastOneArray<WeatherType>`

DayKev committed:
Implement `AtLeastOneOf<TerrainType>`

DayKev committed:
Merge branch 'beta' into refactor/weather-terrain-funcs

DayKev committed:
Merge branch 'beta' into refactor/weather-terrain-funcs

DayKev committed:
Fix merge issue

DayKev committed:
Merge branch 'beta' into refactor/weather-terrain-funcs

flx-sta commented on src/data/battler-tags/weather-highest-stat-boost-tag.ts:
```diff
@@ -24,6 +25,6 @@ export class WeatherHighestStatBoostTag extends HighestStatBoostTag implements W
   override loadTag(source: BattlerTag | any): void {
     super.loadTag(source);
-    this.weatherTypes = source.weatherTypes.map((w) => w as WeatherType);
+    (this as Mutable<this>).weatherTypes = source.weatherTypes.map((w) => w as WeatherType);
```
Q: Why is `as Mutable<this>` necessary?

DayKev replied:
The `weatherTypes` field is `readonly` but when loading tags from save data we bypass the constructor.

flx-sta commented on src/data/mystery-encounters/mystery-encounter-requirements.ts:
```diff
@@ -314,7 +314,8 @@ export class WeatherRequirement extends EncounterSceneRequirement {
     let token = "";
     if (currentWeather !== WeatherType.NONE) {
-      token = WeatherType[currentWeather].replace("_", " ").toLocaleLowerCase();
+      // TODO: don't reinvent the wheel
```
This todo still relevant?

DayKev replied:
Yeah, there's a PR that consolidates/simplifies/removes redundancy in the handling of strings at PKR that needs to be ported over at some point, which will resolve the "TODO".
cf https://github.com/pagefaultgames/pokerogue/pull/6112 + https://github.com/pagefaultgames/pokerogue/pull/6161/files

flx-sta commented on src/data/terrain.ts:
```diff
@@ -186,3 +184,9 @@ export function getTerrainBlockMessage(pokemon: Pokemon, terrainType: TerrainTyp
   });
 }
+
+export function getTerrainAnim(terrainType: TerrainType): CommonAnim {
```
NIT: docs.

Exported functions should always have one

DayKev replied:
Whoops, missed this one.

flx-sta commented on src/enums/weather-type.ts:
```diff
@@ -1,21 +1,31 @@
-  SUNNY,
+import type { ObjectValues } from "#types/utility-types";
+
+export const WeatherType = {
```
NIT: docs

DayKev replied:
The enum members already have docs.

flx-sta commented on src/enums/weather-type.ts:
```diff
@@ -1,21 +1,31 @@
+  STRONG_WINDS: 9,
+} as const;
+
+export type WeatherType = ObjectValues<typeof WeatherType>;
```
NIT: docs

flx-sta commented on src/ui/components/arena-flyout.ts:
```diff
@@ -42,14 +42,14 @@ interface ArenaEffectInfo {
-function getFieldEffectText(arenaTagType: string): string {
-  if (!arenaTagType) {
+function getFieldEffectText(flyoutDisplayText: string): string {
+  if (!flyoutDisplayText || flyoutDisplayText.toLowerCase() === "none") {
```
NIT: Magic string "none" should be a constant

DayKev replied:
Added a comment, I couldn't think of a succinct variable name that explained everything.

DayKev committed:
Add docs to `getTerrainAnim`

DayKev committed:
Add comment to `getFieldEffectText`

DayKev committed:
Merge branch 'beta' into refactor/weather-terrain-funcs

DayKev committed:
Merge branch 'beta' into refactor/weather-terrain-funcs

innerthunder commented on src/@types/utility-types.ts:
```diff
@@ -67,6 +67,9 @@ export type Mutable<T> = {
 };
 
+/** Requires an array to contain at least one item. */
+export type AtLeastOneArray<T> = [T, ...T[]];
```
nit: I'd prefer `AtLeastOne` or `AtLeastOneOf`. `AtLeastOneArray` is a bit verbose and could be misconstrued as "at least one array of T" instead of "array of at least one T".

DayKev replied:
This specifically only works with arrays though. There's [other](https://github.com/pagefaultgames/pokerogue/blob/bcc79a6118d1be91eaecc1ec51a4bdf46532a69e/src/%40types/helpers/type-helpers.ts#L102-L109) `AtLeastOne` types that I don't want to overlap with.
I could move "array" to the beginning so it's `ArrayAtLeastOne`, or maybe instead go with `AtLeastOneItem`?

innerthunder replied:
How about `NonEmptyArray`?

DayKev replied:
Yeah, that sounds good.

DayKev committed:
Merge branch 'beta' into refactor/weather-terrain-funcs

DayKev committed:
Resolve merge issues

DayKev committed:
Update `Arena#tryOverrideWeather` doc

DayKev committed:
Rename `AtLeastOneArray` to `NonEmptyArray`
